File: translator_advanced.py
import asyncio
import logging
import re
import time
from typing import Dict, Optional, Tuple
from deep_translator import GoogleTranslator
from langdetect import detect, LangDetectException
import emoji

class AdvancedTranslationService:
    """Улучшенный сервис перевода с поддержкой сленга и контекста"""
    
    def __init__(self):
        self.logger = logging.getLogger(__name__)
        self.cache: Dict[str, str] = {}
        self.cache_size_limit = 1000
        self.stats = {
            'translations': 0,
            'cache_hits': 0,
            'errors': 0,
            'avg_time': 0.0
        }
        
        # Инициализация переводчиков
        self.ru_en_translator = GoogleTranslator(source='ru', target='en')
        self.en_ru_translator = GoogleTranslator(source='en', target='ru')
        
        self.logger.info("AdvancedTranslationService инициализирован")
        
    def preprocess_text(self, text: str) -> str:
        """Предобработка текста для лучшего перевода"""
        if not text or not text.strip():
            return text
            
        # Нормализуем пробелы
        text = re.sub(r'\s+', ' ', text.strip())
        
        return text
        
    def postprocess_text(self, text: str, target_lang: str) -> str:
        """Постобработка переведенного текста"""
        if not text:
            return text
            
        # Исправляем типичные ошибки перевода
        if target_lang == 'ru':
            # Исправляем "ты" на "вы" в формальном контексте
            text = re.sub(r'\bты\b', 'вы', text, flags=re.IGNORECASE)
            # Исправляем "ваш" на "твой" в неформальном контексте
            text = re.sub(r'\bваш\b', 'твой', text, flags=re.IGNORECASE)
            
        elif target_lang == 'en':
            # Исправляем формальные формы на неформальные для чата
            text = re.sub(r'\bI am\b', "I'm", text, flags=re.IGNORECASE)
            text = re.sub(r'\byou are\b', "you're", text, flags=re.IGNORECASE)
            text = re.sub(r'\bwe are\b', "we're", text, flags=re.IGNORECASE)
            text = re.sub(r'\bthey are\b', "they're", text, flags=re.IGNORECASE)
            
        return text
        
    def detect_language(self, text: str) -> Optional[str]:
        """Определяет язык текста"""
        if not text or not text.strip():
            return None
            
        try:
            # Убираем эмодзи и специальные символы для определения языка
            clean_text = re.sub(r'[^\w\s]', '', text)
            if len(clean_text) < 3:
                return None
                
            lang = detect(clean_text)
            return lang
            
        except LangDetectException:
            # Fallback: проверяем наличие кириллицы
            if re.search(r'[а-яё]', text, re.IGNORECASE):
                return 'ru'
            elif re.search(r'[a-z]', text, re.IGNORECASE):
                return 'en'
            return None
            
    def translate_text(self, text: str, target_lang: str) -> Optional[str]:
        """Переводит текст с улучшенной обработкой"""
        if not text or not text.strip():
            return text
            
        start_time = time.time()
        
        try:
            # Проверяем кэш
            cache_key = f"{text}_{target_lang}"
            if cache_key in self.cache:
                self.stats['cache_hits'] += 1
                return self.cache[cache_key]
                
            # Предобработка
            processed_text = self.preprocess_text(text)
            
            # Определяем исходный язык
            source_lang = self.detect_language(processed_text)
            if not source_lang:
                self.logger.warning(f"Не удалось определить язык для: {text}")
                return None
                
            # Если язык уже целевой, возвращаем как есть
            if source_lang == target_lang:
                return text
                
            # Выбираем переводчик
            if source_lang == 'ru' and target_lang == 'en':
                translator = self.ru_en_translator
            elif source_lang == 'en' and target_lang == 'ru':
                translator = self.en_ru_translator
            else:
                # Fallback для других языков
                translator = GoogleTranslator(source=source_lang, target=target_lang)
                
            # Выполняем перевод
            translated = translator.translate(processed_text)
            
            # Постобработка
            result = self.postprocess_text(translated, target_lang)
            
            # Кэшируем результат
            if len(self.cache) >= self.cache_size_limit:
                # Удаляем старые записи
                old_keys = list(self.cache.keys())[:100]
                for key in old_keys:
                    del self.cache[key]
                    
            self.cache[cache_key] = result
            
            # Обновляем статистику
            self.stats['translations'] += 1
            translation_time = time.time() - start_time
            self.stats['avg_time'] = (
                (self.stats['avg_time'] * (self.stats['translations'] - 1) + translation_time) 
                / self.stats['translations']
            )
            
            self.logger.info(f"Перевод: {text} -> {result} (время: {translation_time:.2f}с)")
            return result
            
        except Exception as e:
            self.stats['errors'] += 1
            self.logger.error(f"Ошибка перевода '{text}': {e}")
            return None
            
    def translate_to_russian(self, text: str) -> Optional[str]:
        """Переводит текст на русский язык"""
        return self.translate_text(text, 'ru')
        
    def translate_to_english(self, text: str) -> Optional[str]:
        """Переводит текст на английский язык"""
        return self.translate_text(text, 'en')
        
    def get_stats(self) -> Dict:
        """Возвращает статистику переводов"""
        return {
            **self.stats,
            'cache_size': len(self.cache),
            'cache_hit_rate': (
                self.stats['cache_hits'] / max(self.stats['translations'], 1) * 100
            )
        }
        
    def clear_cache(self):
        """Очищает кэш переводов"""
        self.cache.clear()
        self.logger.info("Кэш переводов очищен")

File: src/bot/translator_advanced.js
const translate = require('google-translate-api-x');
const EventEmitter = require('events');

class AdvancedTranslationService extends EventEmitter {
    constructor() {
        super();
        
        this.cache = new Map();
        this.maxCacheSize = 1000; // Максимальный размер кэша
        this.maxCacheAge = 1000 * 60 * 60; // 1 час в миллисекундах
        this.cacheCleanupInterval = null;
        
        // Rate limiting
        this.requestQueue = [];
        this.isProcessingQueue = false;
        this.maxRequestsPerMinute = 30; // Лимит запросов в минуту
        this.requestTimes = [];
        
        this.stats = {
            translations: 0,
            cacheHits: 0,
            errors: 0,
            totalTime: 0,
            cacheEvictions: 0
        };
        
        // Запускаем периодическую очистку кэша
        this.startCacheCleanup();
        
        this.log('Продвинутый переводчик инициализирован');
    }

    log(message) {
        console.log(`[Translator] ${message}`);
    }

    preprocessText(text) {
        if (!text || typeof text !== 'string') return '';
        
        // Normalize text
        let processed = text.trim();
        
        // Handle emojis and special characters
        processed = processed.replace(/[\u{1F600}-\u{1F64F}]|[\u{1F300}-\u{1F5FF}]|[\u{1F680}-\u{1F6FF}]|[\u{1F1E0}-\u{1F1FF}]|[\u{2600}-\u{26FF}]|[\u{2700}-\u{27BF}]/gu, '');
        
        // Remove excessive whitespace
        processed = processed.replace(/\s+/g, ' ');
        
        return processed;
    }

    postprocessText(text, targetLang) {
        if (!text) return text;
        
        let processed = text;
        
        // Fix common formal/informal issues
        if (targetLang === 'ru') {
            // Make Russian translation more informal for chat
            processed = processed.replace(/^Привет, /, 'Привет, ');
            processed = processed.replace(/^Здравствуйте, /, 'Привет, ');
        } else if (targetLang === 'en') {
            // Make English translation more informal for chat
            processed = processed.replace(/^Hello, /, 'Hey, ');
            processed = processed.replace(/^Good morning, /, 'Hey, ');
        }
        
        return processed;
    }

    detectLanguage(text) {
        if (!text || text.length < 3) return null;
        
        const latinChars = (text.match(/[a-zA-Z]/g) || []).length;
        const cyrillicChars = (text.match(/[а-яё]/gi) || []).length;
        const totalChars = latinChars + cyrillicChars;
        
        if (totalChars === 0) return null;
        
        const latinPercent = (latinChars / totalChars) * 100;
        const cyrillicPercent = (cyrillicChars / totalChars) * 100;
        
        if (cyrillicPercent > 50) return 'ru';
        if (latinPercent > 50) return 'en';
        
        return null; // Mixed language
    }

    async translateToRussian(text) {
        return this.translate(text, 'en', 'ru');
    }

    async translateToEnglish(text) {
        return this.translate(text, 'ru', 'en');
    }

    async translate(text, sourceLang, targetLang) {
        const startTime = Date.now();
        
        try {
            // Preprocess text
            const processedText = this.preprocessText(text);
            if (!processedText) {
                this.stats.errors++;
                return null;
            }
            
            // Check cache
            const cacheKey = `${processedText}_${sourceLang}_${targetLang}`;
            const cachedValue = this.getCacheValue(cacheKey);
            if (cachedValue) {
                this.stats.cacheHits++;
                this.log(`Cache hit for: ${processedText.substring(0, 50)}...`);
                return cachedValue;
            }
            
            // Detect language if not provided
            const detectedLang = sourceLang || this.detectLanguage(processedText);
            if (!detectedLang) {
                this.log(`Could not detect language for: ${processedText}`);
                this.stats.errors++;
                return null;
            }
            
            // Perform translation with rate limiting
            this.log(`Translating: ${processedText.substring(0, 50)}... (${detectedLang} -> ${targetLang})`);
            
            const result = await this.performTranslationWithRateLimit(processedText, detectedLang, targetLang);
            
            if (!result || !result.text) {
                this.log(`Translation failed for: ${processedText}`);
                this.stats.errors++;
                return null;
            }
            
            // Postprocess result
            const finalResult = this.postprocessText(result.text, targetLang);
            
            // Cache result with timestamp
            this.setCacheWithTimestamp(cacheKey, finalResult);
            
            // Update stats
            this.stats.translations++;
            this.stats.totalTime += Date.now() - startTime;
            
            this.log(`Translation successful: ${processedText.substring(0, 30)}... -> ${finalResult.substring(0, 30)}...`);
            
            return finalResult;
            
        } catch (error) {
            this.log(`Translation error: ${error.message}`);
            this.stats.errors++;
            return null;
        }
    }

    getStats() {
        const avgTime = this.stats.translations > 0 ? this.stats.totalTime / this.stats.translations : 0;
        return {
            ...this.stats,
            averageTime: avgTime,
            cacheSize: this.cache.size,
            cacheHitRate: this.stats.translations > 0 ? (this.stats.cacheHits / this.stats.translations) * 100 : 0
        };
    }

    setCacheWithTimestamp(key, value) {
        // Проверяем размер кэша и очищаем при необходимости
        if (this.cache.size >= this.maxCacheSize) {
            this.evictOldestEntries();
        }
        
        this.cache.set(key, {
            value: value,
            timestamp: Date.now()
        });
    }
    
    getCacheValue(key) {
        const entry = this.cache.get(key);
        if (!entry) return null;
        
        // Проверяем возраст записи
        if (Date.now() - entry.timestamp > this.maxCacheAge) {
            this.cache.delete(key);
            this.stats.cacheEvictions++;
            return null;
        }
        
        return entry.value;
    }
    
    evictOldestEntries() {
        const entriesToDelete = Math.floor(this.maxCacheSize * 0.2); // Удаляем 20% самых старых
        const entries = Array.from(this.cache.entries());
        
        // Сортируем по времени создания
        entries.sort((a, b) => a[1].timestamp - b[1].timestamp);
        
        for (let i = 0; i < entriesToDelete && i < entries.length; i++) {
            this.cache.delete(entries[i][0]);
            this.stats.cacheEvictions++;
        }
        
        this.log(`Evicted ${entriesToDelete} old cache entries`);
    }
    
    startCacheCleanup() {
        // Очистка кэша каждые 10 минут
        this.cacheCleanupInterval = setInterval(() => {
            this.cleanupExpiredEntries();
        }, 1000 * 60 * 10);
    }
    
    cleanupExpiredEntries() {
        const now = Date.now();
        let expiredCount = 0;
        
        for (const [key, entry] of this.cache.entries()) {
            if (now - entry.timestamp > this.maxCacheAge) {
                this.cache.delete(key);
                expiredCount++;
            }
        }
        
        if (expiredCount > 0) {
            this.stats.cacheEvictions += expiredCount;
            this.log(`Cleaned up ${expiredCount} expired cache entries`);
        }
    }
    
    stopCacheCleanup() {
        if (this.cacheCleanupInterval) {
            clearInterval(this.cacheCleanupInterval);
            this.cacheCleanupInterval = null;
        }
    }
    
    async performTranslationWithRateLimit(text, fromLang, toLang) {
        return new Promise((resolve, reject) => {
            this.requestQueue.push({
                text,
                fromLang,
                toLang,
                resolve,
                reject,
                timestamp: Date.now()
            });
            
            this.processQueue();
        });
    }
    
    async processQueue() {
        if (this.isProcessingQueue || this.requestQueue.length === 0) {
            return;
        }
        
        this.isProcessingQueue = true;
        
        try {
            while (this.requestQueue.length > 0) {
                // Проверяем лимит запросов
                if (!this.canMakeRequest()) {
                    const waitTime = this.getWaitTime();
                    this.log(`Rate limit reached, waiting ${waitTime}ms`);
                    await this.sleep(waitTime);
                    continue;
                }
                
                const request = this.requestQueue.shift();
                
                // Проверяем, не устарел ли запрос (более 30 секунд)
                const requestAge = Date.now() - request.timestamp;
                if (requestAge > 30000) {
                    this.log(`Skipping stale translation request (${requestAge}ms old)`);
                    request.reject(new Error('Request timeout'));
                    continue;
                }
                
                try {
                    // Записываем время запроса
                    this.requestTimes.push(Date.now());
                    
                    // Выполняем перевод
                    const result = await translate(request.text, {
                        from: request.fromLang,
                        to: request.toLang
                    });
                    
                    request.resolve(result);
                    
                } catch (error) {
                    this.log(`Translation API error: ${error.message}`);
                    this.stats.errors++;
                    
                    // Если это ошибка сети или API, пробуем повторить
                    if (error.message.includes('network') || error.message.includes('timeout')) {
                        this.log(`Retrying translation due to network error...`);
                        // Возвращаем запрос в очередь для повторной попытки
                        this.requestQueue.unshift(request);
                    } else {
                        request.reject(error);
                    }
                }
                
                // Небольшая задержка между запросами
                await this.sleep(100);
            }
        } catch (error) {
            this.log(`Queue processing error: ${error.message}`);
        } finally {
            this.isProcessingQueue = false;
        }
    }
    
    canMakeRequest() {
        const now = Date.now();
        const oneMinuteAgo = now - 60000;
        
        // Очищаем старые записи
        this.requestTimes = this.requestTimes.filter(time => time > oneMinuteAgo);
        
        return this.requestTimes.length < this.maxRequestsPerMinute;
    }
    
    getWaitTime() {
        if (this.requestTimes.length === 0) return 0;
        
        const oldestRequest = Math.min(...this.requestTimes);
        const timeUntilOldestExpires = 60000 - (Date.now() - oldestRequest);
        
        return Math.max(timeUntilOldestExpires, 1000); // Минимум 1 секунда
    }
    
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
    
    clearCache() {
        this.cache.clear();
        this.stats.cacheEvictions += this.cache.size;
        this.log('Translation cache cleared');
    }
}

module.exports = { AdvancedTranslationService };
